src/miner.cpp

```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
+        if (!inBlock.count(parent))
+        {
+            return true;
+        }
     }
-    assert(nBlockSize == 80);  // BU block header is always 80 bytes
+    return false;
+}
 
+bool BlockAssembler::TestForBlock(CTxMemPool::txiter iter)
+{
+    uint64_t nTxSize = iter->GetTxSize();
 
-    unsigned int nCoinbaseSize=0;
-    // Compute coinbase transaction WITHOUT FEES just to get its size.  We will recompute this at the end when we know the fees.
+    if (nBlockSize + nTxSize > nBlockMaxSize)
     {
-      txNew.vout[0].nValue = 0;  // Will be fixed below, but we need to adjust the size for a possible 9 byte varint
-      txNew.vin[0].scriptSig = CScript() << ((int) 0) << CScriptNum(0);  // block height will be fixed below, but we need to adjust the size for a possible 9 byte varint
-
-      // BU005 add block size settings to the coinbase
-      std::string cbmsg = FormatCoinbaseMessage(BUComments, minerComment);
-      const char* cbcstr = cbmsg.c_str();
-      vector<unsigned char> vec(cbcstr, cbcstr+cbmsg.size());
-      COINBASE_FLAGS = CScript() << vec;
-      // Chop off any extra data in the COINBASE_FLAGS so the sig does not exceed the max.  
-      // we can do this because the coinbase is not a "real" script...
-      if (txNew.vin[0].scriptSig.size() + COINBASE_FLAGS.size() > MAX_COINBASE_SCRIPTSIG_SIZE)
+        // If the block is so close to full that no more txs will fit
+        // or if we've tried more than 50 times to fill remaining space
+        // then flag that the block is finished
+        if (nBlockSize > nBlockMaxSize - 100 || lastFewTxs > 50)
         {
-          COINBASE_FLAGS.resize(MAX_COINBASE_SCRIPTSIG_SIZE - txNew.vin[0].scriptSig.size());
+            blockFinished = true;
+            return false;
         }
-      txNew.vin[0].scriptSig = txNew.vin[0].scriptSig + COINBASE_FLAGS;
-      nCoinbaseSize = txNew.GetSerializeSize(SER_NETWORK, PROTOCOL_VERSION) + 16;  // This code serialized the transaction but the 2 zeros above (nValue and block height) got encoded into 2 bytes, yet these are varints so there is a possible 16 more bytes 
-      // BU005 END
+        // Once we're within 1000 bytes of a full block, only look at 50 more txs
+        // to try to fill the remaining space.
+        if (nBlockSize > nBlockMaxSize - 1000)
+        {
+            lastFewTxs++;
+        }
+        return false;
     }
-    
-    nBlockSize += std::max(nCoinbaseSize,(unsigned int) coinbaseReserve.value);  // BU Miners take the block we give them, wipe away our coinbase and add their own.  So if their reserve choice is bigger then our coinbase then use that.
-    
-    uint64_t nBlockTx = 0;
-    unsigned int nBlockSigOps = 100;  // Reserve 100 sigops for miners to use in their coinbase transaction
-    int lastFewTxs = 0;
-    CAmount nFees = 0;
 
+    uint64_t nTxSigOps = iter->GetSigOpCount();
+    if (nBlockSize + nTxSize <= BLOCKSTREAM_CORE_MAX_BLOCK_SIZE) // Enforce the "old" sigops for <= 1MB blocks
```

Comment:
The actual block size is not an indication of the rules, so this is incorrect, just like above.

The fact if there has been a >1MB block on the chain before is the only indication.